### PR TITLE
Do not empty attributes if they are not provided when user profile is enabled

### LINKED ISSUE #11096: Do not empty attributes if they are not provided when user profile is enabled
### Describe the bug

When the user declarative profile feature is enabled, attributes that are defined in the User Profile are always present in the user attributes even if they are not set (Admin Panel -> Users -> user -> attibutes tab). 

Also, on an attempt to update a user through PUT endpoint attributes are removed from the user.

### Context
We are using ABAC model and make use of the Keycloak user attributes extensively. To prevent users from accessing areas they should not have access to by modifying their attributes we enabled the Declarative User Profile feature and disabled write access to all the important attributes for the users leavin read access only (via Admin Panel -> Realm settings -> User Profile tab -> Attributes subtab ->  attribute -> Permissions section -> Can user edit = FALSE,  Can user view = TRUE).
At the same time, administrators should have read and write access for those attributes (via Admin Panel -> Realm settings -> User Profile tab -> Attributes subtab ->  attribute -> Permissions section -> Can admin edit = TRUE,  Can admin view = TRUE). Attributes are not required in our case.

#### Problem 1
But when this is configured in this way, administrators can see attributes with empty values in the user attributes page (Admin Panel -> Users -> user -> Attributes tab) even if the attribute has not been set for the user. This is confusing for the administrators.

#### Problem 2
We make use of the Admin REST API to make some modifications on the users, for example, update email. But to make it happen it's necessary to send the whole user representation with all the attributes otherwise they will be removed. This makes it necessary to do GET user request first to get full user representation modify it by updating email address and send PUT request to make an update. This introduces unneccessary HTTP call that could be avoided (and can be avoided in case Declarative User Profile feature is disabled)

### Version

16.1.1

### Expected behavior

1. Attributes that are not set for a user are not shown in the Admin Panel -> Users -> user -> Attributes tab
2. On a request to the `PUT /auth/admin/realms/{realm}/users/{id}` endpoint without `attributes` field in the request body all existing user attributes must preserve as it works when User Declarative Profile Feature is disabled.

### Actual behavior

1. For users without attributes from the User Profile configuration fields are still shown on the user attributes tab in the Admin panel with an empty value
2. On a request to the `PUT /auth/admin/realms/{realm}/users/{id}` endpoint without `attributes` field in the request body all existing user attributes are removed when User Declarative Profile Feature is disabled.

### How to Reproduce?

1. Enable Declarative User Profile feature https://www.keycloak.org/docs/latest/server_admin/index.html#enabling-the-user-profile
2. Create a user with username `test`, email `test@test.com`, and **without any attributes**
3. Go to Admin Panel -> Realm settings -> User Profile tab -> Attributes subtab ->  click Create 
4. Create a new attribute with the following parameters and click save:
    a. name: test-attribute
    b. Required: FALSE
    c. Permissions -> Can admin edit: TRUE
    d. Permissions -> Can admin view = TRUE
    e. Permissions -> Can user edit: FALSE
    f. Permissions -> Can user view = TRUE    
5. Go to the user created on the Step 2 Admin Panel -> Users -> user with username `test` (from Step 2) -> Attributes tab
    a. `test-attribute` is shown with empty value even though it's not set for the user. (Actual behavior 1)
6. Set user attribute `test-attribute` value to `Test Value` and save the user
7. Make a PUT request to the `PUT /auth/admin/realms/{realm}/users/{id}` endpoint to update user email for the user created on the step 2. Payload: 
```json
{
    "email": "new-email@test.com"
}
```
8. Check user Admin Panel -> Users -> user with username `test` (from Step 2)
    a. User email has been updated to the `new-email@test.com` (expected), however, 
    b. on the User Attributes tab `test-attribute` value is empty (Actual behavior 2) even though it was not present in the PUT request. And this won't happen in case Declarative User Profile feature is not enabled

### Anything else?

_No response_

### LINKED ISSUE #13550: Preview Feature "Update Email" clears first and last name upon email confirmation
### Describe the bug

When the new UPDATE_EMAIL feature is enabled, going through the process of changing your email address clears the first and last name fields of your user profile.

### Version

19.0.1

### Expected behavior

Changing the email address should not affect any other profile fields.

### Actual behavior

After clicking the confirmation link in the email, the email address gets successfully updated (as expected), but at the same time the first and last name are cleared.

### How to Reproduce?

* Enable the UPDATE_EMAIL feature (e.g. KC_FEATURES=update-email)
* Create a realm
  * which has the "Update Email" required action enabled
  * which has email verification enabled
* Create a user with an email address, first and last name
* Log in as that user and go to the account console. Observe the email address, first and last name in the profile information.
* Click "Update email" and enter a new (different) email adress
* Click the confirmation link sent to that email adress
* Go to the account console again. The email address has been updated, but the first and last name are gone.

### Anything else?

_No response_

### LINKED ISSUE #13561: New Update Email flow not working when "User Profile Enabled" enabled from realm.
### Describe the bug

When "User Profile Enabled" is disabled and update email is enabled , first name and last name became empty on clicking the email validation link. It clears the user profile attributes which were not sent.

But once we enable the "User Profile Enabled" from Realm General setting, we are getting error while confirming new email id on update.

image

### Version

V19.0.1

### Expected behavior

Only email id should be updated. It should work fine with "User Profile Enabled" enabled or disabled.

### Actual behavior

With "User Profile Enabled" disabled, first name and last name became empty.

With "User Profile Enabled" enabled, getting error "Please specify this field"

### How to Reproduce?

From the realm disable "User Profile Enabled" and enable "verify email". Go to user account profile and update email address. On clicking email confirmation link, first name and last name is removed from profile.

From the realm enable "User Profile Enabled" and enable "verify email".  Go to user account profile and update email address, getting error "Please specify this field"

### Anything else?

Only email id should be updated and no other use profile fields should be checked and updated

### PR DESCRIPTION
**Pull request by pedroigor**
Closes #11096
Closes #13550
Closes #13561

### CODE REVIEW

**Comment by pedroigor** [comment]
@ioemat Would you mind reviewing this one?

**Comment by ioemat** [comment]
> @ioemat Would you mind reviewing this one?

sure, I'll take a look later today

**Comment by ioemat** [comment]
@pedroigor just a few small comments

**Comment by pedroigor** [comment]
@ioemat Where are they ? :)

**Comment by ioemat** [comment]
> @ioemat Where are they ? :)

In the code, see above

**Comment by ioemat** [comment]
I did some tests and found one more thing:

If there is an attribute defined that is marked as _required_ and you try to update the user without passing any attributes (e.g. just update username like in `{ "firstName": "fist",   "lastName": "last"}`}, you  receive an error (`"errorMessage": "Please specify attribute attr1."`).

Since attributes are now preserved in such a case, this error doesn't make sense anymore - the provided fields should be changed and attributes should be left untouched.

**Comment by pedroigor** [comment]
The comments you did are not showing to me ...

If you can elaborate more about the required field issue, I appreciate it. Are you talking about a specific context (e.g.: account) ?

**Comment by ioemat** [comment]
> The comments you did are not showing to me ...

sorry, yes, I still had to confirm by review ....  [-x, you should see it now

> 
> If you can elaborate more about the required field issue, I appreciate it. Are you talking about a specific context (e.g.: account) ?

Here's what I did:

* configure a realm `foo` with user profile switched on
* add configuration for a required attribute `attr1`:
```javascript
...
   {
      "selector": {
        "scopes": []
      },
      "permissions": {
        "view": [
          "user",
          "admin"
        ],
        "edit": [
          "user",
          "admin"
        ]
      },
      "name": "attr1",
      "required": {
        "roles": [],
        "scopes": []
      }
    }
  ]
}
```
* add a user `user1` with  `attr1`=`value1`
* `GET http://localhost:8080/admin/realms/foo/users/1f332a2a-f907-4a9a-8be4-8b8af6c717f2 ` returns the user document
```javascript
{
    "id": "1f332a2a-f907-4a9a-8be4-8b8af6c717f2",
    "createdTimestamp": 1662360555815,
    "username": "user1",
    "enabled": true,
    "totp": false,
    "emailVerified": false,
    "attributes": {
        "attr1": [
            "value1"
        ]
    },
    "disableableCredentialTypes": [],
    "requiredActions": [],
    "notBefore": 0,
    "access": {
        "manageGroupMembership": true,
        "view": true,
        "mapRoles": true,
        "impersonate": true,
        "manage": true
    }
}
```
* Try to set `firstName` and `lastName` for this user (without passing the attributes)
    * `PUT http://localhost:8080/admin/realms/foo/users/1f332a2a-f907-4a9a-8be4-8b8af6c717f2`
```javascript
{
    "id": "1f332a2a-f907-4a9a-8be4-8b8af6c717f2",
    "firstName": "John",
    "lastName" : "Doe"
}
```
* An error is returned (http status code 400 - Bad request), document: 
```javascript
{ "errorMessage": "Please specify attribute attr1." }
```

When doing the same with `attr1` being optional, it works as expected:  `firstName` and `lastName`are updated by the `PUT` while the attribute values are kept. I'd expect the same behavior when there are required attributes.

File: server-spi-private/src/main/java/org/keycloak/userprofile/AttributeMetadata.java [hunk 1]
```diff
@@ -175,16 +175,12 @@ public AttributeMetadata addValidator(List<AttributeValidatorMetadata> validator
             this.validators = new ArrayList<>();
         }
 
+        this.validators.removeIf(validators::contains);
```
[the remark this answers is not in the document]
Comment (pedroigor) [reply server-spi-private/src/main/java/org/keycloak/userprofile/AttributeMetadata.java:178]: This code is responsible to replace existing validators if they are being set.

File: server-spi-private/src/main/java/org/keycloak/userprofile/DefaultUserProfile.java [hunk 2]
```diff
@@ -115,6 +115,9 @@ private UserModel updateInternal(UserModel user, boolean removeAttributes, Attri
                 List<String> updatedValue = attribute.getValue().stream().filter(Objects::nonNull).collect(Collectors.toList());
 
                 if (!CollectionUtil.collectionEquals(currentValue, updatedValue)) {
+                    if (!removeAttributes && updatedValue.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (pedroigor) [reply server-spi-private/src/main/java/org/keycloak/userprofile/DefaultUserProfile.java:118]: The behavior is the same when the declarative user profile is disabled. We don't want to update the value of attributes if the list is not provided.

**Comment by pedroigor** [comment]
> When doing the same with `attr1` being optional, it works as expected: `firstName` and `lastName`are updated by the `PUT` while the attribute values are kept. I'd expect the same behavior when there are required attributes.

I'm not following ... If the attribute is marked as required the `400` status code is correct. You can't update the user if the required attributes are missing.

**Comment by pedroigor** [comment]
@ioemat As discussed, the validation during updates was improved to include existing user attributes when the attribute map is not sent in the payload.

**Review by ioemat — APPROVED** [review]
@pedroigor thanks for considering my comments, LGTM now :)

**Review by vmuzikar — APPROVED** [review]

**Pull request merged**